pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,278 @@
+type PolicyRuleWithSubject struct {
+	// Subject is the user, group or serviceaccount that this rule cares about.
+	// +optional
+	Subject rbac.Subject
+	// Rule is the target resource or the subresource the rule cares about. APIGroups, Resources, etc.
+	// Required.
+	Rule rbac.PolicyRule
+}
+
+// FlowSchemaStatus represents the current state of a flow-schema.
+type FlowSchemaStatus struct {
+	// Current state of flow-schema.
+	Conditions []FlowSchemaCondition
+}
+
+// FlowSchemaCondition describes conditions for a flow-schema.
+type FlowSchemaCondition struct {
+	// Type is the type of the condition.
+	Type FlowSchemaConditionType
+	// Status is the status of the condition.
+	// Can be True, False, Unknown.
+	Status api.ConditionStatus
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time
+	// Unique, one-word, CamelCase reason for the condition's last transition.
+	Reason string
+	// Human-readable message indicating details about last transition.
+	Message string
+}
+
+// FlowSchemaConditionType is a valid value for FlowSchemaStatusCondition.Type
+type FlowSchemaConditionType string
+
+// FlowSchemaConditionStatus is a valid value for FlowSchemaStatusCondition.Status
+type FlowSchemaConditionStatus string
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfiguration represents the configuration of a priority level.
+type PriorityLevelConfiguration struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of a request-priority.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec PriorityLevelConfigurationSpec
+	// Current status of a request-priority.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status PriorityLevelConfigurationStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfigurationList is a list of PriorityLevelConfiguration objects.
+type PriorityLevelConfigurationList struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+	// Items is a list of request-priorities.
+	Items []PriorityLevelConfiguration
+}
+
+// PriorityLevelConfigurationSpec is specification of a priority level
+type PriorityLevelConfigurationSpec struct {
+	// GlobalDefault specifies whether this priority level should be considered as the default priority for requests
+	// that do not have any priority level. Only one PriorityClass can be marked as `globalDefault`.
```

Comment:
s/can/should/ .  We have no way to _enforce_ "can".

Reply:
I would expect, if multiple, one of them should get a rejection notice in its status.

Consider requiring the name of the global default to be "default" and vice versa; then it is in fact enforcable.

Reply:
@lavalamp : are you thinking of setup and transition scenarios?

Reply:
> I would expect, if multiple, one of them should get a rejection notice in its status.

so there'll be only one actively working default, and the other stucking e.g. w/ a `Established=False` condition in the status. what if the working one steps down? pick a random one from the rejected list, or it's sth we don't need to care in the alpha stage?

Reply:
I interpreted the sentence as "at most one"; I think we need to supply (hard-code) a default if one isn't currently set in etcd.

Reply:
Sorry, it didn't show me Min's response-- My preference is to require the name be set to "default", so that there can only ever be one.

Reply:
@lavalamp: please review https://github.com/kubernetes/enhancements/blob/7a144ac413b0d9b3fbe43c0bc5dd843e2e7c19d7/keps/sig-api-machinery/20190228-priority-and-fairness.md#default-behavior if you have not already.  I wrote that without (a) knowing that we have a way to ensure the existence of API objects before the server goes into business (is this really true?  Where are examples?) and (b) being sure we wanted to mandate a particular set of API objects as an initial load.  And what if the operator wants a variant of our standard initial load?

I dislike using "default" as a name because it is semantically fragile.  Remember that one man's application is another man's platform, and both might be using kube API objects.  A small fragment of an example appears at the end of https://github.com/kubernetes/enhancements/blob/7a144ac413b0d9b3fbe43c0bc5dd843e2e7c19d7/keps/sig-api-machinery/20190228-priority-and-fairness.md#example-configuration .  A kube cluster can be used for multiple things, and some of those are platforms for still other things.

Reply:
@yue9944882 : my thinking was that if there are multiple PriorityLevelConfiguration objects with `GlobalDefault == true` this would normally only be during a transition period when an administrator is adjusting configuration, and in this case (1) picking any of them is a reasonable choice and (2) feedback on which is picked is not necessary.  Also, remember that we are talking about some logic that runs independently in each apiserver; I would prefer to not ask for feedback from all the apiservers, particularly about uninteresting details of a transition.

Reply:
Remember also that this design includes metrics that _do_ provide feedback about what is happening.

Reply:
> My preference is to require the name be set to "default", so that there can only ever be one.

aye, i think will solve the problem. 

> (1) picking any of them is a reasonable choice and (2) feedback on which is picked is not necessary. 

i guess that will drive the users angry. the default configuration is equally critical in the system. or even, the most users will be working on the default configuration if the users are getting lazy to cusomize their own configuration from the default one. we have to clearly know which configuration and how that is the working one at all time.